src/components/devsupport/components/measure/measure.component.tsx

```diff
@@ -4,75 +4,78 @@
- * but `force` property may be used to measure any time it's needed.
- * DON'T USE THIS FLAG IF THE COMPONENT RENDERS FIRST TIME OR YOU KNOW `onLayout` WILL BE CALLED.
- */
-export const MeasureElement = (props: MeasureElementProps): MeasuringElement => {
-
-  const ref = React.useRef<any>();
-
-  const bindToWindow = (frame: Frame, window: Frame): Frame => {
-    if (frame.origin.x < window.size.width) {
-      return frame;
-    }
-
-    const boundFrame: Frame = new Frame(
-      frame.origin.x - window.size.width,
-      frame.origin.y,
-      frame.size.width,
-      frame.size.height,
-    );
-
-    return bindToWindow(boundFrame, window);
-  };
-
-  const onUIManagerMeasure = (x: number, y: number, w: number, h: number): void => {
-    const frame: Frame = bindToWindow(new Frame(x, y, w, h), Frame.window());
-    props.onMeasure(frame);
-  };
-
-  const measureSelf = (): void => {
-    const node: number = findNodeHandle(ref.current);
-    UIManager.measureInWindow(node, onUIManagerMeasure);
-  };
-
-  if (props.force) {
-    measureSelf();
-  }
-
-  return React.cloneElement(props.children, { ref, onLayout: measureSelf });
-};
+ import React from 'react';
+ import {
+   findNodeHandle,
+   UIManager,
+   StatusBar,
+ } from 'react-native';
+ import { Frame } from './type';
+ import { ModalService } from '../../../theme/modal/modal.service';
+ 
+ export interface MeasureElementProps<P = any> {
+   force?: boolean;
+   onMeasure: (frame: Frame) => void;
+   children: React.ReactElement<P>;
+ }
+ 
+ export type MeasuringElement<P = any> = React.ReactElement;
+ 
+ /**
+  * Measures child element size and it's screen position asynchronously.
+  * Returns measure result in `onMeasure` callback.
+  *
+  * Usage:
+  *
+  * ```tsx
+  * const onMeasure = (frame: Frame): void => {
+  *   const { x, y } = frame.origin;
+  *   const { width, height } = frame.size;
+  *   ...
+  * };
+  *
+  * <MeasureElement onMeasure={onMeasure}>
+  *   <ElementToMeasure />
+  * </MeasureElement>
+  * ```
+  *
+  * By default, it measures each time onLayout is called,
+  * but `force` property may be used to measure any time it's needed.
+  * DON'T USE THIS FLAG IF THE COMPONENT RENDERS FIRST TIME OR YOU KNOW `onLayout` WILL BE CALLED.
+  */
+ export const MeasureElement = (props: MeasureElementProps): MeasuringElement => {
+ 
+   const ref = React.useRef<any>();
+ 
+   const bindToWindow = (frame: Frame, window: Frame): Frame => {
+     if (frame.origin.x < window.size.width) {
+       return frame;
+     }
+ 
+     const boundFrame: Frame = new Frame(
+       frame.origin.x - window.size.width,
+       frame.origin.y,
+       frame.size.width,
+       frame.size.height,
+     );
+ 
+     return bindToWindow(boundFrame, window);
+   };
+ 
+   const onUIManagerMeasure = (x: number, y: number, w: number, h: number): void => {
+     const originY = ModalService.getTopInsetsState ? y + StatusBar.currentHeight || 0 : y;
```

Comment:
Is it possible not import modal service here, and instead just pass a prop?

Reply:
done. But the prop needs to be passed to each `<MeasureElement />`. As the props needs to be mandatory - I was thinking it's independent enough to just import it.